R/PipeOpTrafo.R

```diff
@@ -0,0 +1,240 @@
+#' @title PipeOpInvertiblePreproc
+#'
+#' @usage NULL
+#' @format Abstract [`R6Class`] inheriting from [`PipeOp`].
+#'
+#' @description
+#' #FIXME:
+#'
+#' @section Construction:
+#' #FIXME:
+#' ```
+#' PipeOpInvertiblePreproc$new(id)
+#' ```
+#'
+#' * `id` :: `character(1)`\cr
+#'   Identifier of resulting object. See `$id` slot of [`PipeOp`].
+#'
+#' @section Input and Output Channels:
+#' #FIXME:
+#'
+#' @section State:
+#' #FIXME:
+#'
+#' @section Internals:
+#' #FIXME:
+#'
+#' @section Methods:
+#' Methods inherited from [`PipeOp`], as well as:
+#' #FIXME:
+#'
+#' @family mlr3pipelines backend related
+#' @family PipeOps
+#' @include PipeOp.R
+#' @export
+PipeOpInvertiblePreproc = R6Class("PipeOpInvertiblePreproc",
+  inherit = PipeOp,
+  public = list(
+    initialize = function(id) {
+      super$initialize(id = id,
+        input = data.table(name = "input", train = "Task", predict = "Task"),
+        output = data.table(name = c("fun", "output"), train = c("NULL", "Task"), predict = c("function", "Task"))
+      )
+    },
+
+    train_internal = function(inputs) {
+      intask = inputs[[1]]$clone(deep = TRUE)
+      intask = self$train_target(intask)
+      list(NULL, intask)
+    },
+
+    predict_internal = function(inputs) {
+      predict_phase_control = self$train_invert(inputs[[1L]])
+      list(
+        fun = private$invert_help(predict_phase_control),
+        task = private$set_target_to_NA_and_possibly_convert_task_type(inputs[[1L]])
+      )
+    },
+
+    train_target = function(task) {
+      # return a modified task and set the state
+      # make use of private$update_target
+      # therefore, typically calculate new_target (name it appropriately), do task$cbind(new_target)
+      # and provide names(new_target) as a character to update_target, e.g.:
+      #
+      # self$state = list()
+      # new_target = foo
+      # names(new_target) = "foo"
+      # task$cbind(new_target)
+      # private$update_target(task, new_target = names(new_target), new_type = NULL, ...)
+      stop("Abstract.")
+    },
+
+    train_invert = function(task) {
+      # return a predict_phase_control object (can be anything)
+      stop("Abstract.")
+    },
+
+    inverter = function(prediction, predict_phase_control) {
+      # function that inverts the predictions and returns a Prediction object
+      stop("Abstract.")
+    }
+  ),
+  private = list(
+    update_target = function(task, new_target, new_type = NULL, ...) {
+      type = if (is.null(new_type)) task$task_type else assert_subset(new_type, mlr_reflections$task_types$type, empty.ok = FALSE)
+      get(mlr_reflections$task_types[type]$task)$new(id = task$id, backend = task$backend, target = new_target, ...)
```

Comment:
See `convert_target` from hurdle models branch;  do not convert task if task_type stays the same.